File: backend/src/main/java/com/autocare/mechanic/service/MechanicService.java
package com.autocare.mechanic.service;

import com.autocare.mechanic.dto.MechanicRequestDTO;
import com.autocare.mechanic.dto.MechanicResponseDTO;
import com.autocare.mechanic.entity.Mechanic;
import com.autocare.mechanic.repository.MechanicRepository;
import com.autocare.shared.exception.BusinessException;
import com.autocare.shared.exception.ResourceNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class MechanicService {

    private final MechanicRepository mechanicRepository;

    @Transactional
    public MechanicResponseDTO create(MechanicRequestDTO request) {
        if (request.getUserId() != null && mechanicRepository.findByUserId(request.getUserId()).isPresent()) {
            throw new BusinessException("Este usuário já está vinculado a outro perfil de mecânico");
        }

        Mechanic mechanic = Mechanic.builder()
                .name(request.getName())
                .specialty(request.getSpecialty())
                .phone(request.getPhone())
                .userId(request.getUserId())
                .isAvailable(true)
                .isActive(true)
                .build();

        mechanic = mechanicRepository.save(mechanic);
        return toResponseDTO(mechanic);
    }

    @Transactional
    public MechanicResponseDTO update(String id, MechanicRequestDTO request) {
        Mechanic mechanic = findById(id);

        if (request.getUserId() != null
                && !request.getUserId().equals(mechanic.getUserId())
                && mechanicRepository.findByUserId(request.getUserId()).isPresent()) {
            throw new BusinessException("Este usuário já está vinculado a outro perfil de mecânico");
        }

        mechanic.setName(request.getName());
        mechanic.setSpecialty(request.getSpecialty());
        mechanic.setPhone(request.getPhone());
        mechanic.setUserId(request.getUserId());

        mechanic = mechanicRepository.save(mechanic);
        return toResponseDTO(mechanic);
    }

    @Transactional
    public void delete(String id) {
        Mechanic mechanic = findById(id);
        mechanic.setActive(false);
        mechanicRepository.save(mechanic);
    }

    @Transactional
    public void setAvailability(String id, boolean available) {
        Mechanic mechanic = findById(id);
        mechanic.setAvailable(available);
        mechanicRepository.save(mechanic);
    }

    public MechanicResponseDTO findByIdResponse(String id) {
        return toResponseDTO(findById(id));
    }

    public List<MechanicResponseDTO> findAll() {
        return mechanicRepository.findByIsActiveTrue()
                .stream()
                .map(this::toResponseDTO)
                .collect(Collectors.toList());
    }

    public List<MechanicResponseDTO> findAvailable() {
        return mechanicRepository.findByIsAvailableTrue()
                .stream()
                .filter(Mechanic::isActive)
                .map(this::toResponseDTO)
                .collect(Collectors.toList());
    }

    public Mechanic findById(String id) {
        return mechanicRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Mecânico não encontrado com ID: " + id));
    }

    private MechanicResponseDTO toResponseDTO(Mechanic mechanic) {
        return MechanicResponseDTO.builder()
                .id(mechanic.getId())
                .name(mechanic.getName())
                .specialty(mechanic.getSpecialty())
                .phone(mechanic.getPhone())
                .isAvailable(mechanic.isAvailable())
                .createdAt(mechanic.getCreatedAt())
                .isActive(mechanic.isActive())
                .build();
    }
}

File: backend/src/mechanic/mechanic.service.ts
import { Injectable } from "@nestjs/common";
import { PrismaService } from "../prisma/prisma.service";
import { BusinessException, ResourceNotFoundException } from "../common/exceptions";
import { MechanicRequestDto } from "./dto/mechanic.dto";

@Injectable()
export class MechanicService {
  constructor(private prisma: PrismaService) {}

  async create(dto: MechanicRequestDto) {
    if (dto.userId) {
      const linked = await this.prisma.mechanic.findUnique({ where: { userId: dto.userId } });
      if (linked) {
        throw new BusinessException("Este usuário já está vinculado a outro perfil de mecânico");
      }
    }

    const mechanic = await this.prisma.mechanic.create({
      data: {
        name: dto.name,
        specialty: dto.specialty,
        phone: dto.phone,
        userId: dto.userId,
        isAvailable: true,
        isActive: true,
      },
    });
    return this.toResponse(mechanic);
  }

  async update(id: string, dto: MechanicRequestDto) {
    const mechanic = await this.findById(id);

    if (dto.userId && dto.userId !== mechanic.userId) {
      const linked = await this.prisma.mechanic.findUnique({ where: { userId: dto.userId } });
      if (linked) {
        throw new BusinessException("Este usuário já está vinculado a outro perfil de mecânico");
      }
    }

    const updated = await this.prisma.mechanic.update({
      where: { id },
      data: { name: dto.name, specialty: dto.specialty, phone: dto.phone, userId: dto.userId },
    });
    return this.toResponse(updated);
  }

  async delete(id: string) {
    await this.findById(id);
    await this.prisma.mechanic.update({ where: { id }, data: { isActive: false } });
  }

  async setAvailability(id: string, available: boolean) {
    await this.findById(id);
    await this.prisma.mechanic.update({ where: { id }, data: { isAvailable: available } });
  }

  async findByIdResponse(id: string) {
    return this.toResponse(await this.findById(id));
  }

  async findAll() {
    const mechanics = await this.prisma.mechanic.findMany({
      where: { isActive: true },
      orderBy: { createdAt: "desc" },
    });
    return mechanics.map((m) => this.toResponse(m));
  }

  async findAvailable() {
    const mechanics = await this.prisma.mechanic.findMany({
      where: { isAvailable: true, isActive: true },
    });
    return mechanics.map((m) => this.toResponse(m));
  }

  // Retorna a entidade crua — usado internamente por ServiceOrderService/BudgetService.
  async findById(id: string) {
    const mechanic = await this.prisma.mechanic.findUnique({ where: { id } });
    if (!mechanic) {
      throw new ResourceNotFoundException(`Mecânico não encontrado com ID: ${id}`);
    }
    return mechanic;
  }

  async findByUserId(userId: string) {
    return this.prisma.mechanic.findUnique({ where: { userId } });
  }

  private toResponse(mechanic: any) {
    return {
      id: mechanic.id,
      name: mechanic.name,
      specialty: mechanic.specialty,
      phone: mechanic.phone,
      isAvailable: mechanic.isAvailable,
      createdAt: mechanic.createdAt,
      isActive: mechanic.isActive,
    };
  }
}
